src/hamiltonian.jl

```diff
@@ -1,8 +1,29 @@
 # TODO: add a type for kinetic energy
 
+# The constructor of `LogDensityValue` will check numerical errors in
+#   `value` and `gradient`.  That is `is_valid` will be performed automatically.
+struct LogDensityValue{Tv<:AbstractFloat, Tg<:AbstractVector{Tv}}
+    value::Tv    # Cached value, e.g. logπ(θ).
+    gradient::Tg # Cached gradient, e.g. ∇logπ(θ).
+end
+
+struct LogDensityFunction{Tf<:Function}
```

Comment:
Will this work? I think `f` and `∇f` will have different types.

Reply:
I'm not sure I understand the issue here. Can you clarify a bit here?

Reply:
See minimal example below:

```julia
julia> struct A{T<:Function}
       v1::T
       v2::T
       end

julia> f() = 1
f (generic function with 1 method)

julia> A(f, f)
A{typeof(f)}(f, f)

julia> g() = :b
g (generic function with 1 method)

julia> A(f, g)
ERROR: MethodError: no method matching A(::typeof(f), ::typeof(g))
Closest candidates are:
  A(::T<:Function, ::T<:Function) where T<:Function at REPL[1]:2
Stacktrace:
 [1] top-level scope at none:0
```